[WIP] `async` and `.await` chapter (#3909)
rust-lang/book: src/ch17-03-more-futures.md

> ## Working With More Than Two Futures

Author: I think this section heading would be more accurate as something like:

```suggestion
## Joining Any Number of Futures
```

Because I think it could be argued that in listing 17-12, we were "working with" a whole bunch of futures (every `.await`, over each loop, plus the outermost one). And at the end of the previous section, we _did_ join 3 futures, so we've already done more than two...

Reviewer: Ah, good call. I also ended up folding together sections 17.03 and 17.04 (after pulling out the deeper-dive material for a new section near the end of the chapter), so I am (for now!) going with “Working With Any Number of Futures” here since it also now includes `race` etc.

When we switched from using two futures to three in the previous section, we

> also had to switch from using `join` to using `join3`. It would be annoying to
> do this every time we changed our code. Happily, we have a macro form of `join`

Author: I think "do this" and "changed our code" could be a tiny bit more specific, something like:

```suggestion
also had to switch from using `join` to using `join3`. It would be annoying to
have to call a different function every time we changed the number of futures we wanted to join. Happily, we have a macro form of `join`
```

Reviewer: Oh yeah, that is *way* clearer. 👍🏼

to which we can pass an arbitrary number of arguments. It also handles awaiting
the futures itself. Thus, we could rewrite the code from Listing 17-12 to use
`join!` instead of `join3`, as in Listing 17-13:

<Listing number="17-13" caption="Using `join!` to wait for multiple futures" file-name="src/main.rs">

```rust
{{#rustdoc_include ../listings/ch17-async-await/listing-17-13/src/main.rs:here}}
```

</Listing>

> This is definitely a nice improvement over needing to swap between `join` and

Author: I'm not sure we _need_ to show `join` and `join3'-- the reader has been using `println!` and (should) know that macros are for taking a variable number of arguments. Any particular reason not to start out using the macro? Then at this point, talk about the common pattern?

Reviewer: Huh! I don’t think I had ever actually framed it as “macros are for taking a variable number of arguments” before reading this comment from you! Interesting. I am definitely open to revisiting the order. I think I defaulted *away* from using `join!` because of how bad a taste `tokio::select!` left in my mouth, but `futures::join!` is much, *much* less magical, so I don’t have any objection to doing that. Will leave this one for a further pass, though, after getting through all the other stuff, since it will require some rejiggering of listings etc.

`join3` and `join4` and so on! However, even this macro form only works when we

> know the number of futures ahead of time. In real-world Rust, though, pushing

Reviewer: Took me a *while* to see the best way to do this, but I did: I actually integrated it into a question at the end of the section! Thanks for the suggestion!

replace `join!` with `join_all`.

<Listing  number="17-14" caption="Storing anonymous futures in a vector and calling `join_all`">

```rust,ignore,does_not_compile
{{#rustdoc_include ../listings/ch17-async-await/listing-17-14/src/main.rs:here}}
```

</Listing>

Unfortunately, this does not compile. Instead, we get this error:

<!-- manual-regeneration
cd listings/ch17-async-await/listing-17-14/
cargo build
copy just the compiler error, and *add* the following text (correctly aligned),
to match the nicer version we will have starting in 1.81

```
   = note: no two async blocks, even if identical, have the same type
   = help: consider pinning your async block and and casting it to a trait object
```

Once 1.81 lands, we can remove the last part of that, since it will include the
message correctly automatically; but we  still need to do the rest of the manual
regeneration, unfortunately.

-->

<!--
TODO: delete this note once 1.81 is out and we update the version note at the
front of the book.
-->

> Note: Beta readers, the error version shown here is landing in Rust 1.81.0!
> If you are using an earlier version, you will see a *much* less helpful error

> > message here. We fixed it as part of writing this chapter!

Author: 🎉

dive into a little more of how the `Future` type actually works, in particular
the idea of *pinning*.

### Pinning and the Pin and Unpin Traits

<!-- TODO: get a *very* careful technical review of this section! -->

Let’s look again at the definition of `Future`, focusing now on its `poll`
method’s `self` type:

```rust
pub trait Future {
    type Output;

    // Required method
    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output>;
}
```

This is the first time we have seen a method where `self` has a type annotation
like this. When we specify the type of `self` like this, we are telling Rust
what type `self` must be to call this method. These kinds of type annotations
for `self` are similar to those for other function parameters, but with the
restriction that the type annotation has to be the type on which the method is
implemented, or a reference or smart pointer to that type. We will see more on
this syntax in Chapter 18. For now, it is enough to know that if we want to poll
a future (to check whether it is `Pending` or `Ready(Output)`), we need a
mutable reference to the type, which is wrapped in a `Pin`.

> `Pin` is a smart pointer, much like `Box`, `Rc`, and the others we saw in

Author: @jamesmunns can you clarify what the difference is between `Pin` and smart pointers (just the heap allocation or?) and what you would call `Pin` instead? A wrapper type?

> `Pin` is a smart pointer, much like `Box`, `Rc`, and the others we saw in

Reviewer: Yeah, I just checked and the standard library docs for `std::pin` and `std::pin::Pin` use “wrapper” for `Pin` itself and “wrapped pointer” for the things it wraps. I had seen it referred to as a “smart pointer” in a number of places, and it stuck in my mind as a result, but I am not sure where, and in any case “wrapper type” makes perfectly good sense—and even matches how the docs describe `Vec<T>`, `Box<T>`, etc. in `std::pin`; the choice there seems to have been to use “wrapper type” as the more general term (into which `Pin<T>` would fit) and “smart pointer” as the more specific one for `Box`, `Rc`, etc. 👍🏼

> `Pin` is a smart pointer, much like `Box`, `Rc`, and the others we saw in

Reviewer 2: Yeah, Id agree with wrapper type, or Board's recent post compared it to a type state. Pin doesn't *do* anything, whereas Rc/Arc/Box are smart to me because they have drop impls that may automatically free the data.

> `Pin` is a smart pointer, much like `Box`, `Rc`, and the others we saw in

Reviewer: Rewrote this to describe `Pin` as a “wrapper type” and to do a bit more compare-and-contrast with the smart pointer types so people understand that it is purely a tool for the compiler to keep track of the guarantees, rather than something with behavior like drop or reference counting.

starts to explain the error message we saw above—but that was in terms of
`Unpin`, not `Pin`. So what exactly are `Pin` and `Unpin`, how do they relate,
and why does `Future` need `self` to be in a `Pin` type to call `poll`?

In [“What Are Futures”][what-are-futures], we described how a series of await
points in a future get compiled into a state machine—and noted how the compiler
helps make sure that state machine follows all of Rust’s normal rules around
safety, including borrowing and ownership. To make that work, Rust looks at what
data is needed between each await point and the next await point or the end of
the async block. It then creates a corresponding variant in the state machine it
creates. Each variant gets the access it needs to the data that will be used in
that section of the source code, whether by taking ownership of that data or by
getting a mutable or immutable reference to it.

So far so good: if we get anything wrong about the ownership or references in a
given async block, the borrow checker will tell us. When we want to move around
the future that corresponds to that block—like moving it into a `Vec` to pass to

> `join_all`—things get trickier.

Author: I think around this spot it would be awesome if we had an example of, say, trying to use a reference in an async block, show the error message, and show how to fix it-- I get that sort of error message alllll the time when i'm switching between sync and async code because it looks like i'm just within a block or something and it should be totally fine to use a reference and then error message and then i have to remember to clone the arc or whatever.

Reviewer: @carols10cents I am going to leave this specific comment open but without making a specific change related to it, because I think the overall suggestion here is a good one but I need to mull on how it works with, and where it belongs in, the chapter after the restructuring I did. It does *not* flow well with this paragraph where it landed in the new section dedicated to discussing the traits, which is (in its current incarnation, anyway) much more abstract. But I think that it probably *needs* exactly this kind of thing to help it *not* be too abstract. I am going to thus table that for the moment and leave this as a note to return to after hitting the rest of this PR’s comments and reworking the opening example.

When we move a future—whether by pushing into a data structure to use as an
iterator with `join_all`, or returning them from a function—that actually means
moving the state machine Rust creates for us. And unlike most other types in
Rust, the futures Rust creates for async blocks can end up with references to
themselves in the fields of any given variant. Any object which has a reference
to itself is unsafe to move, though, because references always point to the
actual memory address of the thing they refer to. If you move the data structure
itself, you *have* to update any references to it, or they will be left pointing

> to the old location.

Reviewer 3: This is really hard to imagine in the abstract. A concrete example would help (and break up the wall of text).

In principle, you could make the Rust compiler try to update every reference to
an object every time it gets moved. That would  potentially be a lot of
performance overhead, especially given there can be a whole web of references
that need updating. On the other hand, if we could make sure the data structure
in question *does not move in memory*, we do not have to update any references.
And this is exactly what Rust’s borrow checker already guarantees: you cannot
move an item which has any active references to it using safe code.

`Pin` builds on that to give us the exact guarantee we need. When we *pin* a
value by wrapping a pointer to it in `Pin`, it can no longer move. Thus, if you
have `Pin<Box<SomeType>>`, you actually pin the `SomeType` value, *not* the
`Box` pointer. In fact, the pinned box pointer can move around freely. Remember:
we care about making sure the data ultimately being referenced stays in its
place. If a pointer moves around, but the data it points to is in the same

> place, there is no problem.

Author: This could use a diagram, along the lines of some of the ones in chapter 4

Reviewer: Ah, very good call; I will make a diagram for this! (Note also that this whole section is now part of a new 17.05, which pulls together all the walk-throughs of the relevant traits.)

moved.

Listing 17-17 shows how we put this all into practice. First, we update the type
annotation for `futures`, with a `Pin` wrapping each `Box`. Second, we use
`Box::pin` to pin the futures themselves.

<Listing number="17-17" caption="Using `Pin` and `Box::pin` to make the `Vec` type check" file-name="src/main.rs">

```rust
{{#rustdoc_include ../listings/ch17-async-await/listing-17-17/src/main.rs:here}}
```

</Listing>

If we compile and run this, we finally get the output we hoped for:

<!-- Not extracting output because changes to this output aren't significant;
the changes are likely to be due to the threads running differently rather than
changes in the compiler -->

```text
received 'hi'
received 'more'
received 'from'
received 'messages'
received 'the'
received 'for'
received 'future'
received 'you'
```

Phew!

There is a bit more we can explore here. For one thing, using `Pin<Box<T>>`
comes with a small amount of extra overhead from putting these futures on the
heap with `Box`—and we are only doing that to get the types to line up. We don’t
actually *need* the heap allocation, after all: these futures are local to this

> particular function. As noted above, `Pin` is itself a smart pointer, so we can

Reviewer 2: Another note re: "smart pointer"

otherwise Rust will still not know to interpret these as dynamic trait objects,
which is what we need them to be in the `Vec`. We therefore `pin!` each future
when we define it, and define `futures` as a `Vec` containing pinned mutable
references to the dynamic `Future` type, as in Listing 17-18.

<Listing number="17-18" caption="Using `Pin` directly with the `pin!` macro to avoid unnecessary heap allocations" file-name="src/main.rs">

```rust
{{#rustdoc_include ../listings/ch17-async-await/listing-17-18/src/main.rs:here}}
```

</Listing>

There is another, more serious, issue as well. We got this far by ignoring the
fact that we might have different `Output` types. For example, in Listing 17-19,
the anonymous future for `a` implements `Future<Output = u32>`, the anonymous
future for `b` implements `Future<Output = &str>`, and the anonymous future for
`c` implements `Future<Output = bool>`.

<Listing number="17-19" caption="Three futures with distinct types" file-name="src/main.rs">

```rust
{{#rustdoc_include ../listings/ch17-async-await/listing-17-19/src/main.rs:here}}
```

</Listing>

We can use `trpl::join!` to await them, because it allows you to pass in

> multiple future types and produces a tuple of those htypes. We *cannot* use

Reviewer 2: ```suggestion
multiple future types and produces a tuple of those types. We *cannot* use
```

`trpl::join_all`, because it requires the futures passed in all to have the same
type. (Remember, that error is what got us started on this adventure with
`Pin`!)

This is a fundamental tradeoff: we can either deal with a dynamic number of
futures with `join_all`, as long as they all have the same type, or we can deal
with a set number of futures with the `join` functions or the `join!` macro,
even if they have different types. This is the same as working with any other
types in Rust, though. Futures are not special, even though we have some nice
syntax for working with them, and that is a good thing.

In practice, you will usually work directly with `async` and `.await`, and
secondarily with functions and macros like `join` or `join_all`. You will only
need to reach for `pin` now and again to use them with those APIs. `Pin` and
`Unpin` are mostly important for building lower-level libraries, or when you are
building a runtime itself, rather than for day to day Rust code. When you see

> them, though, now you will know what to do!

Author: So this paragraph points to a potential reordering -- perhaps the `pin` discussion should be later in the chapter than this, we should just show `join`/`join_all` and discuss the different scenarios for using them here. The later section should have `Pin` in the section title, and be introduced with something like "if you're interested in the details on `Pin` now, here you go, if not, feel free to move on to chapter 18 and return here if you get an error message talking about `Pin` and `Unpin`"...?

Reviewer: Yeah, this is more or less exactly the reordering I ended up doing—albeit at an even larger scale! I kept the discussion of `pin!` here, but I basically said “just follow the compiler’s advice” and noted that we’ll discuss the “why” a bit later in the chapter.
